streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -989,16 +994,18 @@ private void to(final TopicNameExtractor<K, V> topicExtractor,
             null,
             optimizableRepartitionNodeBuilder);
 
-        final OptimizableRepartitionNode<K, V> optimizableRepartitionNode = optimizableRepartitionNodeBuilder.build();
-        builder.addGraphNode(streamsGraphNode, optimizableRepartitionNode);
+        if (repartitionNode == null || !name.equals(repartitionName)) {
```

Comment:
This is the fix.  We'll generate the `repartitionNode` if it's the first time through or if the user has provided a name for the repartition topic.

Reply:
Hmmm... I am wondering if just bumping the index would be sufficient and the optimizer would merge the node automatically?

I am also not sure about the code structure: so far, the DSL layer does not know much about optimizations (even if we "leak" a little bit into it, as we built up the `StreamsGraphNode` graph... We would push some optimization decisions into the DSL layer thus spreading out "optimization code"? On the other hand, just inserting one `OptimizableRepartitionNode` is much more efficient than inserting multiple and let the optimizer remove them later?

I am also wondering, if we could do the same for other repartition topics? 

Last question: this method is also use for stream-table joins and thus, if one joins a stream with two tables, would this change be backward incompatible? Or would two stream-table joins fail with the same `InvalidTopologyException`?

Reply:
>Hmmm... I am wondering if just bumping the index would be sufficient and the optimizer would merge the node automatically?

I hadn't thought of that, but it should work.  I initially had concerns for topology compatibility,  but I don't think that is the case since users can't create re-use a `KStream` node in joins that needs repartitioning.

>I am also not sure about the code structure: so far, the DSL layer does not know much about optimizations (even if we "leak" a little bit into it, as we built up the StreamsGraphNode graph... We would push some optimization decisions into the DSL layer thus spreading out "optimization code"? On the other hand, just inserting one OptimizableRepartitionNode is much more efficient than inserting multiple and let the optimizer remove them later?

Yeah, I agree the current approach is leaking too much optimization into the current code.  I think it will be better to just go ahead and create the topology "as is" and let the optimizer do its job.

>I am also wondering, if we could do the same for other repartition topics?

We probably should have a consistent approach.  How about I make the changes in this PR for the Join repartition topics (incrementing the index of the repartition node name) and do a follow-on PR to address the other repartition topics?

>Last question: this method is also use for stream-table joins and thus, if one joins a stream with two tables, would this change be backward incompatible? Or would two stream-table joins fail with the same InvalidTopologyException?

I believe two stream-table joins without this fix will fail with the same exception, but I'll add some tests to confirm.

Reply:
Since you made the mistake of asking my opinion, here it is :) :

> bumping the index
It's true that users can't currently reuse the KStream, so there's no compatibility issue there, but we can't bump the index for the _first_ repartition topic, or we would break every topology that uses generated repartition topic names already. So, either way, we have to cache something to tell us to do something different on the "first reuse" (i.e., the second use of the KStream).

Since we have to do that anyway, maybe it's fine to just cache the repartition node itself instead of a flag that says "bump the index next time". 

> leaking optimizations into the DSL

I'm on the fence about whether this is an "optimization" or "reasonable behavior". It sort of feels like the latter, and the only reason we needed to introduce the "repartition-collapsing" optimization is that we failed to introduce reasonable behavior from the beginning. Also, my read is that the DSL builder and the optimizer are not cleanly separated right now anyway, and if we ever want to build more optimizations, we'll most likely need to make another pass on both anyway. We're also starting to think about topology evolution (cc @cadonna ), which makes this a less scary prospect, as we can then implement a mechanism to _compatibly_ introduce new optimizations. In other words, I'm not taking a hard stance, but leaning in the direction of doing the more efficient thing than the more pure thing, since we're not currently super pure anyway.

> Other repartition topics

I think we'd better leave it alone for now, implement topology evolution, then migrate to a completely pure and consistent approach.

Reply:
> We probably should have a consistent approach. How about I make the changes in this PR for the Join repartition topics (incrementing the index of the repartition node name) and do a follow-on PR to address the other repartition topics?

@bbejeck Works for me.

> I'm on the fence about whether this is an "optimization" or "reasonable behavior".

@bbejeck @vvcephei That was my reasoning from my other comments about "don't make the same mistake again", too (cf. https://github.com/apache/kafka/pull/8504#issuecomment-618073161). Atm, we just need to keep the old behavior for backward compatibility reasons and only give "reasonable" behavior via opt-in to the optimization. IMHO, optimization should be the default behavior anyway (not the other way round; we just have it that way due to compatibility constraints) and you should even be able to turn it off (if possible)